### PR TITLE
Add shortcuts for glue resources

### LINKED ISSUE #73: [proposal] Shortcuts for Glue resources
The Glue resources have a lot of properties with pretty poor documentation on what values are allowed, what properties are required, or what those properties are even used for. Often I've had to figure out what I want my CloudFormation to look like by first creating the thing I want via API call or the console, then asking the API to give me the resource description and translating that into CloudFormation.

By making shortcuts, we might be able to short-circuit some of that nonsense and make it easier to manage databases and tables in CloudFormation.

Proposed shortcuts:

- **GlueDatabase**: A name is all that's required.
- **OrcGlueTable**: Define a table sourced from ORC files on S3. Name, Location, Columns are required.
- **JsonGlueTable**: Define a table sourced from line-delimited JSON files on S3. Name, Location, Columns are required.
- **GlueView**: Define a view by providing a name and some SQL (I'm still proving out that this works in CloudFormation).

Curious to hear thoughts and suggestions @Ramshackle-Jamathon @drboyer @danmactough

### PR DESCRIPTION
**Pull request by rclark**
Adds shortcuts:
- GlueDatabase
- GlueTable
- GlueJsonTable
- GlueOrcTable
- GlueView

## Things to review

- Look at the `-defaults.json` fixtures for each of these resource types. Do the defaults I selected look valid? Should any others be added?

- The `GlueTable` resource is more or less a blank slate, and an easy opportunity to provide invalid input. That said, it's the base class for the other resources which are supposed to constrain the use-case and provide "functional" tables out-of-the-box. Do we feel good about this?

- Generally I have not written any more input validation than "these things have to exist" into shortcuts. Do we need more validation in any of these cases in order for the shortcuts to be useful?

## Other TODO

- Finish documentation

cc @danmactough @drboyer @Ramshackle-Jamathon 

This PR fixes #73

### CODE REVIEW

File: lib/shortcuts/glue-orc-table.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+const GlueTable = require('./glue-table');
+
+class GlueOrcTable extends GlueTable {
+  constructor(options = {}) {
+    let { Parameters } = options;
+    const { Location } = options;
+
+    const required = [Location];
+    if (required.some((variable) => !variable))
+      throw new Error('You must provide a Location');
```
Comment (Ramshackle-Jamathon) [inline lib/shortcuts/glue-orc-table.js]: stub leftovers?

File: lib/shortcuts/glue-orc-table.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    const TableType = 'EXTERNAL_TABLE';
+    const InputFormat = 'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat';
+    const OutputFormat = 'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat';
+    const SerdeInfo = {
+      SerializationLibrary: 'org.apache.hadoop.hive.ql.io.orc.OrcSerde',
+      Parameters: { 'serialization.format': '1' }
+    };
+    Parameters = Object.assign(
+      { EXTERNAL: 'TRUE', 'orc.compress': 'SNAPPY' },
```
Comment (Ramshackle-Jamathon) [inline lib/shortcuts/glue-orc-table.js]: Orc files can be uncompressed or use a different compression strategy. Imo this should be left configurable but default to being uncompressed.

Comment (Ramshackle-Jamathon) [reply lib/shortcuts/glue-orc-table.js]: I don't think `EXTERNAL: 'TRUE'` is required

Comment (Ramshackle-Jamathon) [reply lib/shortcuts/glue-orc-table.js]: supported compression formats: https://docs.aws.amazon.com/athena/latest/ug/compression-formats.html

File: lib/shortcuts/glue-orc-table.js (hunk 1 above)
Comment (rclark) [reply lib/shortcuts/glue-orc-table.js]: Nope, this is supposed to be here. I think that's what you're asking?

File: lib/shortcuts/glue-orc-table.js (hunk 2 above)
Comment (rclark) [reply lib/shortcuts/glue-orc-table.js]: I removed default values for any Parameters. They're still accepted if the caller provides them.

File: lib/shortcuts/glue-orc-table.js (hunk 1 above)
Comment (Ramshackle-Jamathon) [reply lib/shortcuts/glue-orc-table.js]: yep 😄 In that case why is Location required?

Comment (rclark) [reply lib/shortcuts/glue-orc-table.js]: Can a table with these input/output formats work as intended without a Location?

Comment (Ramshackle-Jamathon) [reply lib/shortcuts/glue-orc-table.js]: Yes, location denotes the default path where partitions will be stored, but you're free to explicitly register a partition into a different path, even if Location is defined.

Comment (rclark) [reply lib/shortcuts/glue-orc-table.js]: Okay I think I see the behavior. Glue only needs to know where each partition's files are located, and the table-level specification just provides a default. 

If you specify a table-level Location but you don't specify a partition-level Location, where does Glue expect the partition data to reside? For example say my table's Location is `s3://bucket/my/table` and I write a partition to it for `partion_column=value`. Where would I put the ORC data on S3 in order for Glue to find it?

Do you see any benefit to forcing table-level declaration of the location? This makes for more useful metadata.

image

Too opinionated?

Comment (Ramshackle-Jamathon) [reply lib/shortcuts/glue-orc-table.js]: Another thing, glue won't automatically pick up partitions that are written to the correct default location, you still need to either tell it to register that particular partition or to "repair" and pick up missing partitions. So using Location won't save an sdk call.

`${LocationUri}/${Location}/dt={}/hr={}` would be the location for partitions expected by a table with dt and hr partitions.

Im familiar enough with the value of `Location` to want to make it required. I can see some confusion coming from it not being prescriptive, you could end up with `location: "bogus"` in the params and still have a functioning table.

Comment (danmactough) [reply lib/shortcuts/glue-orc-table.js]: > Do you see any benefit to forcing table-level declaration of the location? This makes for more useful metadata that's viewable in the Glue UI.

A lot of these parameters are Hive parameters leaking through the Glue abstraction. This is one of them. `Location` is not required by Hive, so I don't think we should make it required.

File: lib/shortcuts/glue-orc-table.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class GlueOrcTable extends GlueTable {
+  constructor(options = {}) {
+    const { Location } = options;
+
+    const required = [Location];
+    if (required.some((variable) => !variable))
+      throw new Error('You must provide a Location');
+
+    const TableType = 'EXTERNAL_TABLE';
```
Comment (danmactough) [inline lib/shortcuts/glue-orc-table.js]: There's also a `MANAGED_TABLE` type that is valid, right? Should we allow that?

File: test/fixtures/shortcuts/glue-json-table-defaults.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+          "Parameters": {
+            "classification": "json"
+          },
+          "PartitionKeys": [],
+          "TableType": "EXTERNAL_TABLE",
+          "StorageDescriptor": {
+            "Compressed": false,
+            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
+            "Location": "s3://fake/location",
+            "NumberOfBuckets": 0,
```
Comment (danmactough) [inline test/fixtures/shortcuts/glue-json-table-defaults.json]: I would expect `NumberOfBuckets` to be absent in the default case and only be present (and `>0` I guess?) if there are `BucketColumns`.

**Comment by danmactough** [comment]
> * Generally I have not written any more input validation than "these things have to exist" into shortcuts. Do we need more validation in any of these cases in order for the shortcuts to be useful?

I'm really torn on this question. On the one hand, don't want to generate an invalid CloudFormation template. On the other hand, CloudFormation will validate the template, so I don't know if there a huge value-add to trying to provide additional validation.

For example, we _could_ validate that any specified `PartitionKeys` or `SortColumns` or `BucketColumns` are defined among the specified `Columns`.

Similarly, if a user specifies `BucketColumns`, then `NumberOfBuckets` must be specified (and, although not explicit in the AWS docs, presumably `NumberOfBuckets` must be greater than zero, too).

Should we enforce those things here? Or defer to CloudFormation?

I'm leaning toward the latter (so, not adding more validation here) unless we know we can save a user from a foot-gun -- where CF would allow a user to define a thing that won't work properly.

**Comment by Ramshackle-Jamathon** [comment]
@rclark @danmactough 🤔 would it be better to move the json/orc shortcuts into the private library? I think there are a few unanswered questions with how those might work that would better sussed out there before adding back to cloudfriend?

**Comment by rclark** [comment]
> Should we enforce those things here? Or defer to CloudFormation?

What I've found is that cloudformation will pretty much accept anything in these string fields that should really belong to some documented enumeration somewhere. And it won't care if you define column names differently in two parts of the table definition.

We should just aim to build the json/orc tables into ones with default configuration that we know constructs a table that functions.

@Ramshackle-Jamathon per chat, an open-ended "let's just try this out for a while" probably isn't going to definitively answer these questions. Give me a minute to actually audit the few fields that these shortcuts actually default in our internal catalog. It might be immediately obvious that there are a few clear groupings.

If you just want to start using this code, feel free to do a prerelease. Just recognize you might have to revisit that system if we change things up before we merge here.

**Comment by vsmart** [comment]
Per discussion w @Ramshackle-Jamathon, just noting here that there is still interest for this feature. I ran into this today and would appreciate cloudfriend-backed template support.

File: lib/shortcuts/glue-view.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
+'use strict';
+
+const GlueTable = require('./glue-table');
+
+class GlueJsonTable extends GlueTable {
```
Comment (rclark) [inline lib/shortcuts/glue-view.js]:
```suggestion
class GlueView extends GlueTable {
```

File: lib/shortcuts/glue-view.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          ViewOriginalText,
+          ViewExpandedText
+        },
+        options
+      )
+    );
+  }
+}
+
+module.exports = GlueJsonTable;
```
Comment (rclark) [inline lib/shortcuts/glue-view.js]:
```suggestion
module.exports = GlueView;
```

**Pull request merged**